client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java

```diff
@@ -907,8 +908,43 @@ public final void shutdown() {
     @Override
     public final void shutdown(final Handler<AsyncResult<Void>> completionHandler) {
 
-        if (shuttingDown.compareAndSet(Boolean.FALSE, Boolean.TRUE)) {
+        closeConnection(completionHandler, true);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public final void disconnect() {
+
+        disconnect(disconnectResult -> {
+            if (disconnectResult.succeeded()) {
+                LOG.info("successfully disconnected from the server");
+            } else {
+                LOG.error("could not disconnect from the server", disconnectResult.cause());
+            }
+        });
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public final void disconnect(final Handler<AsyncResult<Void>> completionHandler) {
+        closeConnection(completionHandler, false);
+    }
+
+    //-----------------------------------< private methods >---
+
+    private void closeConnection(final Handler<AsyncResult<Void>> completionHandler, final Boolean isShuttingDown) {
```

Comment:
what is `isShuttingDown` good for?

Reply:
to determine whether the `closeConnection` method is called during a `HonoClient#shutdown` call (i.e `isShuttingDown` is true) or `HonoClient#disconnect` call(i.e `isShuttingDown` is false). the `shuttingDown` global flag, once set to true (during shutdown), is never reset back again. this is the reason why a subsequent connect/reconnect always fails. but in the case of the disconnect, once the disconnect is successful we need to reset the flag back (to allow subsequent calls to disconnect again). so `isShuttingDown` determine whether the flag needs to be reset or not